File: Java/118. Pascal's Triangle.java
/**
 * https://leetcode.com/problems/pascals-triangle/
 */

import java.util.ArrayList;
import java.util.List;

public class Solution {
    // O(numRows*numRows) time complexity
    // O(numRows*numRows) space complexity
    public List<List<Integer>> generate(int numRows) {
        List<List<Integer>> res = new ArrayList<>();
        if (numRows == 0)
            return res;
        res.add(new ArrayList<>());
        res.get(0).add(1);

        for (int i = 1; i < numRows; i++) {
            List<Integer> prev = res.get(i - 1);
            List<Integer> curr = new ArrayList<>();
            curr.add(1);
            for (int j = 1; j < i; j++) {
                curr.add(prev.get(j) + prev.get(j - 1));
            }
            curr.add(1);
            res.add(curr);
        }
        return res;
    }
}

File: Kotlin/118. Pascal's Triangle.kt
/**
 * https://leetcode.com/problems/pascals-triangle/
 */

class Solution {
    // O(numRows*numRows) time complexity
    // O(numRows*numRows) space complexity
    fun generate(numRows: Int): List<List<Int>> {
        val res = ArrayList<MutableList<Int>>()

        if (numRows == 0) return res
        res.add(mutableListOf())
        res[0].add(1)

        for (i in 1 until numRows) {
            val prev = res[i - 1]
            val curr = mutableListOf<Int>()
            curr.add(1)
            for (j in 1 until i) {
                curr.add(prev[j - 1] + prev[j])
            }
            curr.add(1)
            res.add(curr)
        }
        return res
    }
}